### DISCUSSION TITLE
Needed Functionalities

### DISCUSSION THREAD

**Discussion by meyerscetbon** [ott-jax/ott#31] [Ideas]
- Continuous discrete Barycenter
- Scaling mechanism for `Geometry` Object, relying for instance on maximal norms of points
- Initialization for `LRSinkhorn` using rank-2 matrices as in https://arxiv.org/abs/2103.04737

**Comment by marcocuturi** [comment]
All done! Thanks for your contributions!

**Comment by marcocuturi** [comment]
Just to give an update on what the plan is in following weeks:
- more implementations of existing Wasserstein barycenter algorithms, notably for generic Riemannian manifolds (as long as a cost operator has a barycenter operation that is coded). As an example of this, maybe consider geometry on the sphere / torus / psd matrices.
- debug the debiased continuous W barycenter (currently this is not correct)
- implementation of a discrete Gromov-Wasserstein barycenter, building e.g. on [this](http://proceedings.mlr.press/v48/peyre16.pdf) for instance. In that process we might have to refresh the discrete barycenter approach.
- implementation of a continuous Gromov-Wasserstein barycenter.
- automatic re-normalization approaches for costs that are selected by default in the solvers.


if you need another functionality for your project, please do not hesitate to add it here!

**Reply by zwei-beiner** [reply]
Hi, are Wasserstein/sinkhorn distances implemented for generic manifolds (in particular, the sphere)? Otherwise, are there any plans to implement this in the near future?

**Reply by marcocuturi** [reply]
OTT's philosophy is (https://ott-jax.readthedocs.io/en/latest/)

> Split geometry from OT solvers in the discrete case: We argue that there should be one, and one implementation only, of every major OT algorithm (Sinkhorn, Gromov-Wasserstein, barycenters, etc…), regardless of the geometric setup that is considered. To give a concrete example, any speedups one may benefit from by using a specific cost (e.g. Sinkhorn being faster when run on a separable cost on histograms supported on a separable grid [[Solomon et al., 2015](https://ott-jax.readthedocs.io/en/latest/bibliography.html#id31)]) should not require a separate reimplementation of a Sinkhorn routine.

If you only need to run sinkhorn on a geometric manifold that has a specific cost, you only need to implement a custom cost function for that manifold, as briefly explored in this [tutorial](https://ott-jax.readthedocs.io/en/latest/tutorials/geometry/000_point_cloud.html#ot-gradient-flows) where a custom cost is defined. There the custom cost is translation invariant (a function of `x-y`) but you could try to implement your own cost directly by simply overriding the `pairwise(x,y)` cost function.
